### PR TITLE
[dynamo] Avoid unnecessary `.detach()` call in `_make_subclass` polyfill

### PR DESCRIPTION
**Pull request by StrongerXi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #151265

This brings down compilation time quite a bit for certain tensor
subclass + `torch.compile` use cases, see #150706.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: torch/_dynamo/polyfills/tensor.py [hunk 1]
```diff
@@ -15,7 +15,10 @@ def make_subclass(
     # suffice for most of Dynamo tracing purposes.
     # https://github.com/pytorch/pytorch/blob/ccfde4dadfa3c342076a1ee387017f84dd4ad2f7/torch/csrc/autograd/python_variable.cpp#L597-L650
     assert len(kwargs) == 0, "_make_subclass only supports requires_grad as keyword arg"
-    data = data.detach()
+
+    # Avoid unnecessary `detach` node in the graph.
+    if data.requires_grad:
+        data = data.detach()
```
Comment (zou3519) [inline torch/_dynamo/polyfills/tensor.py:21]: https://github.com/pytorch/pytorch/blob/4a47dd9b3f5d3aa587a8f909ce53f3a5eddcde6d/torch/csrc/autograd/python_variable.cpp#L617-L618

make_subclass unconditionally applies detach. Why should we avoid doing this when requires_grad is False? I'm not convinced the semantics are the same.

Comment (StrongerXi) [reply torch/_dynamo/polyfills/tensor.py:21]: > Why should we avoid doing this when requires_grad is False?

Mainly because it gets added aggressively into the Dynamo graph and increases AOTDispatcher compilation time _quite_ a bit: #150706, which might be caused by #151142 but I'm not sure.

I also thought about using `alias` by apparently the semantics aren't the same either:
```
>>> import torch
>>> x = torch.ones(2)
>>> x_detach = x.detach()
>>> x_alias = torch.ops.aten.alias(x)
>>> x += 1
>>> x_detach
tensor([2., 2.])
>>> x_alias
tensor([2., 2.])
>>> x_detach._is_view()
False
>>> x_alias._is_view()
True
```

Comment (StrongerXi) [reply torch/_dynamo/polyfills/tensor.py:21]: Do we have to perfectly emulate the [detach semantics](https://github.com/pytorch/pytorch/blob/d5a8bd068821ca0cd036400c13a561452d391b15/aten/src/ATen/native/TensorShape.cpp#L4504-L4512) here though? In fact before I added this `detach` emulation in #149483, we never even emulated it at all.

The detach call under `requires_grad=False` looks mostly like a super shallow copy (not even view), and since the detached data will be fed into a `as_subclass` or `torch.Tensor` call below (which is another shallow copy), I wonder if there's any way for the end user to observe the differences of a perfect `detach `emulation v.s. not in real code

Comment (zou3519) [reply torch/_dynamo/polyfills/tensor.py:21]: I think I am fine as long as the output tensor of the _make_subclass call is not a view of the input tensor (when requires_grad=False). I don't know how to trigger this code, if you have a repro, the view check looks like:
```
x = ... (input)
y = ...  # the make_subclass call
assert y._base is None
```

**Review by zou3519 — COMMENTED** [review]
as I mentioned before, let's check the aliasing behavior. If it is correct then I'm happy to ship this.

**Comment by StrongerXi** [comment]
> as I mentioned before, let's check the aliasing behavior. If it is correct then I'm happy to ship this.

I looked into how Dynamo, `torch.Tensor` and `torch.Tensor._base` interact, turns out there's a long standing latent bug... The takeaways:
1. Eager semantics: `as_subclass` returns a view (i.e., has `_base`), and `_make_subclass` doesn't (despite sharing the same underlying data as the input).
2. For tensor subclasses, dynamo graph breaks on `._base` or `._is_view`. So to match eager semantics, Dynamo just needs to reconstruct the tensor subclass instances properly. Historically we always invoke `as_subclass` on the output tensor of the fx graph, but that only matches the `as_subclass` semantics (i.e., we always return a view): 
https://github.com/pytorch/pytorch/blob/aafe8a67b5ce97eb11e1f6caac172131c667c5eb/torch/_dynamo/codegen.py#L270-L282

To fix (2), i.e., make sure a tensor subclass constructed via `_make_subclass` isn't a view, a tensor subclass VT needs to remember how it was constructed, i.e., via `as_subclass` or `_make_subclass`, and then invoke the appropriate function during reconstruction. I'll come back to this after I finish other issues on my plate.
